src/CpAmm.ts

```diff
@@ -484,19 +475,31 @@ export class CpAmm {
    */
   async getPositionsByUser(user: PublicKey): Promise<
     Array<{
-      publicKey: PublicKey;
-      account: PositionState;
+      positionNftAccount: PublicKey;
+      position: PublicKey;
+      positionState: PositionState;
     }>
   > {
-    const userNfts = await getAllNftByUser(
+    const userPositionAccounts = await getAllUserPositionNftAccount(
       this._program.provider.connection,
       user
     );
-    const positions = await this._program.account.position.all();
-    const result = positions.filter((item) =>
-      userNfts.includes(item.account.nftMint.toString())
+    if (userPositionAccounts.length === 0) {
+      return [];
+    }
+
+    const positionAddresses = userPositionAccounts.map((account) =>
+      derivePositionAddress(account.positionNft, this._program.programId)
+    );
+
+    const positionStates = await this._program.account.position.fetchMultiple(
```

Comment:
what happen if there is an invalid position address? can function return error?

Reply:
Returns null for invalid positions (accounts not found or incorrect discriminator). I've added a filter to remove all null results.